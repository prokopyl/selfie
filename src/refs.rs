//! Reference type stand-ins, to be used with [`Selfie`](crate::Selfie) and
//! [`SelfieMut`](crate::SelfieMut) type declarations.
//!
//! These types are stand-ins that allow to name reference types without naming any associated
//! lifetime. The full type can then be reconstructed with an arbitrary lifetime using the
//! [`RefType`] trait.
//!
//! This is necessary for [`Selfie`](crate::Selfie), as it has to work with a self-referential
//! lifetime, which cannot be explicitly named and has to be reconstructed internally.
//! In essence, this is a roundabout way to achieve Higher-Kinded Polymorphism.
//!
//! This module provides stand-ins for common reference types, but you can create your own by
//! implementing the [`RefType`] trait yourself.

use crate::{Selfie, SelfieMut};
use core::marker::PhantomData;

/// A trait for reference type stand-ins to be combined with an arbitrary lifetime `'a`, to recreate
/// the full reference type.
///
/// # Example
///
/// Implementing [`RefType`] for a custom referential type, so it can be used in a [`Selfie`]:
///
/// ```
/// use std::pin::Pin;
/// use selfie::refs::RefType;
/// use selfie::Selfie;
///
/// #[derive(Copy, Clone)]
/// struct MyReferentialType<'a>(&'a str);
///
/// struct MyReferentialTypeStandIn;
/// impl<'a> RefType<'a> for MyReferentialTypeStandIn {
///     type Ref = MyReferentialType<'a>;
/// }
///
/// // MyReferentialType can now be used in Selfies!
/// let data = Pin::new("Hello, world!".to_owned());
/// let selfie: Selfie<String, MyReferentialTypeStandIn> = Selfie::new(data, |str| MyReferentialType(&str[0..5]));
///
/// assert_eq!("Hello", selfie.referential().0);
/// ```
///
/// Here is a dummy example showing how [`RefType`] stand-ins are used internally:
///
/// ```
/// use selfie::refs::{Ref, RefType};
///
/// // These two type declarations are equivalent
/// const STR_1: &'static str = "Hello, world!";
/// const STR_2: <Ref<str> as RefType<'static>>::Ref = "Hello, world!";
///
/// assert_eq!(STR_1, STR_2);
/// ```
pub trait RefType<'a> {
    /// The full reference type that is to be created when combined with the lifetime `'a`.
    type Ref: 'a + Sized;
}

/// A stand-in for a shared reference `&T`.
///
/// # Example
///
/// ```
/// use std::pin::Pin;
/// use selfie::refs::Ref;
/// use selfie::Selfie;
///
/// let data = Pin::new("Hello, world!".to_owned());
/// let selfie: Selfie<String, Ref<str>> = Selfie::new(data, |str| &str[0..5]);
///
/// assert_eq!("Hello", selfie.referential());
/// ```
pub struct Ref<T: ?Sized>(PhantomData<T>);

impl<'a, T: 'a + ?Sized> RefType<'a> for Ref<T> {
    type Ref = &'a T;
}

/// A stand-in for a mutable reference `&mut T`.
///
/// # Example
///
/// ```
/// use std::pin::Pin;
/// use selfie::refs::Mut;
/// use selfie::SelfieMut;
///
/// let data = Pin::new("Hello, world!".to_owned());
/// let selfie: SelfieMut<String, Mut<str>> =
///     SelfieMut::new(data, |str| &mut Pin::into_inner(str)[0..5]);
///
/// selfie.with_referential(|str| assert_eq!("Hello", *str));
/// ```
pub struct Mut<T: ?Sized>(PhantomData<T>);

impl<'a, T: 'a + ?Sized> RefType<'a> for Mut<T> {
    type Ref = &'a mut T;
}

/// A stand-in for a [`Selfie`](crate::Selfie) holding a reference type as its owned pointer.
///
/// # Example
///
/// ```
/// use std::pin::Pin;
/// use selfie::refs::{Ref, SelfieRef};
/// use selfie::Selfie;
///
/// let data = Pin::new("Hello, world!".to_owned());
/// let selfie: Selfie<String, SelfieRef<Ref<str>, Ref<str>>> = Selfie::new(data, |str| {
///     let substr = Pin::new(&str[0..5]);
///     Selfie::new(substr, |str| &str[3..])
/// });
///
/// assert_eq!("Hello, world!", selfie.owned());
/// selfie.with_referential(|r1| {
///     assert_eq!("Hello", r1.owned());
///     assert_eq!("lo", r1.referential());
/// });
/// ```
pub struct SelfieRef<P, R>(PhantomData<P>, PhantomData<R>)
where
    P: ?Sized,
    R: ?Sized;

impl<'a, P, R> RefType<'a> for SelfieRef<P, R>
where
    P: RefType<'a>,
    R: 'a + for<'this> RefType<'this>,
{
    type Ref = Selfie<'a, P::Ref, R>;
}

/// A stand-in for a [`SelfieMut`](crate::SelfieMut) holding a reference type as its owned pointer.
///
/// # Example
///
/// ```
/// use std::pin::Pin;
/// use selfie::refs::{Mut, SelfieRefMut};
/// use selfie::SelfieMut;
///
/// let data = Pin::new("Hello, world!".to_owned());
/// let selfie: SelfieMut<String, SelfieRefMut<Mut<str>, Mut<str>>> = SelfieMut::new(data, |str| {
///     let substr = Pin::new(&mut Pin::into_inner(str)[0..5]);
///     SelfieMut::new(substr, |str| &mut Pin::into_inner(str)[3..])
/// });
///
/// selfie.with_referential(|inner_selfie| {
///     inner_selfie.with_referential(|str| assert_eq!("lo", *str))
/// });
/// ```
pub struct SelfieRefMut<P, R>(PhantomData<P>, PhantomData<R>)
where
    P: ?Sized,
    R: ?Sized;

impl<'a, P, R> RefType<'a> for SelfieRefMut<P, R>
where
    P: RefType<'a>,
    R: 'a + for<'this> RefType<'this>,
{
    type Ref = SelfieMut<'a, P::Ref, R>;
}

// Other std types

impl<'a, R: RefType<'a>> RefType<'a> for Option<R> {
    type Ref = Option<R::Ref>;
}
