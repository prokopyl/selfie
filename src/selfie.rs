use crate::refs::*;
use crate::utils::*;
use core::fmt::{Debug, Formatter};
use core::ops::DerefMut;
use core::pin::Pin;
use stable_deref_trait::StableDeref;

pub struct Selfie<'a, P: 'a, R: for<'this> RefType<'this> + ?Sized> {
    // SAFETY: enforce drop order!
    referential: <R as RefType<'a>>::Ref,
    owned: Pin<P>,
}

impl<'a, P: StableDeref + 'a, R: for<'this> RefType<'this> + ?Sized> Selfie<'a, P, R> {
    pub fn new(
        owned: Pin<P>,
        handler: for<'this> fn(&'this P::Target) -> <R as RefType<'this>>::Ref,
    ) -> Self
    where
        P::Target: 'a,
    {
        // SAFETY: This type does not expose anything that could expose referential longer than owned exists
        let derefd = unsafe { detach_lifetime(owned.as_ref()) }.get_ref();

        let referential = handler(derefd);

        Self { referential, owned }
    }

    #[inline]
    pub fn owned(&self) -> &P::Target {
        self.owned.as_ref().get_ref()
    }

    #[inline]
    pub fn referential(&self) -> &<R as RefType<'a>>::Ref {
        &self.referential
    }

    #[inline]
    pub fn into_inner(self) -> Pin<P> {
        self.owned
    }
}

impl<'a, P: 'a + StableDeref, R: for<'this> RefType<'this> + ?Sized> Debug for Selfie<'a, P, R>
where
    P::Target: Debug,
    <R as RefType<'a>>::Ref: Debug,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("Selfie")
            .field("owned", &self.owned())
            .field("referential", self.referential())
            .finish()
    }
}

pub struct SelfieMut<'a, P: 'a, R: for<'this> RefType<'this> + ?Sized> {
    // SAFETY: enforce drop order!
    referential: <R as RefType<'a>>::Ref,
    pinned: Pin<P>,
}

impl<'a, P: StableDeref + DerefMut + 'a, R: for<'this> RefType<'this> + ?Sized>
    SelfieMut<'a, P, R>
{
    #[inline]
    pub fn new(
        mut pinned: Pin<P>,
        handler: for<'this> fn(Pin<&'this mut P::Target>) -> <R as RefType<'this>>::Ref,
    ) -> Self
    where
        P::Target: 'a,
    {
        // SAFETY: This type does not expose anything that could expose referential longer than owned exists
        let derefd = unsafe { detach_lifetime_mut(pinned.as_mut()) };

        let referential = handler(derefd);

        Self {
            referential,
            pinned,
        }
    }

    #[inline]
    pub fn referential(&self) -> &<R as RefType<'a>>::Ref {
        &self.referential
    }

    #[inline]
    pub fn referential_mut(&mut self) -> &mut <R as RefType<'a>>::Ref {
        &mut self.referential
    }

    #[inline]
    pub fn into_inner(self) -> Pin<P> {
        self.pinned
    }
}

impl<'a, P: StableDeref + DerefMut + 'a, R: for<'this> RefType<'this> + ?Sized> Debug
    for SelfieMut<'a, P, R>
where
    <R as RefType<'a>>::Ref: Debug,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("SelfieMut")
            .field("referential", self.referential())
            .finish()
    }
}
