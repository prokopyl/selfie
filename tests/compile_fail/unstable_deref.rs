use core::ops::Deref;
use core::pin::Pin;
use selfie::refs::Ref;
use selfie::Selfie;

struct UnstableInt(pub i32);

impl Deref for UnstableInt {
    type Target = i32;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

pub fn main() {
    let int = Pin::new(UnstableInt(42));
    let data: Selfie<UnstableInt, Ref<i32>> = Selfie::new(int, |i| i);

    data.with_referential(|r| assert_eq!(&42, *r));

    let data = Box::new(data);
    data.with_referential(|r| assert_eq!(&42, *r));
}
