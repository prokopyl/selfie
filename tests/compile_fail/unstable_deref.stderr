error[E0277]: the trait bound `UnstableInt: stable_deref_trait::StableDeref` is not satisfied
  --> tests/compile_fail/unstable_deref.rs:18:59
   |
18 |     let data: Selfie<UnstableInt, Ref<i32>> = Selfie::new(int, |i| i);
   |                                               ----------- ^^^ expected an implementor of trait `stable_deref_trait::StableDeref`
   |                                               |
   |                                               required by a bound introduced by this call
   |
note: required by a bound in `Selfie::<'a, P, R>::new`
  --> src/selfie.rs
   |
   | impl<'a, P: StableDeref + 'a, R: for<'this> RefType<'this> + ?Sized> Selfie<'a, P, R> {
   |             ^^^^^^^^^^^ required by this bound in `Selfie::<'a, P, R>::new`
help: consider borrowing here
   |
18 |     let data: Selfie<UnstableInt, Ref<i32>> = Selfie::new(&int, |i| i);
   |                                                           +
18 |     let data: Selfie<UnstableInt, Ref<i32>> = Selfie::new(&mut int, |i| i);
   |                                                           ++++

error[E0599]: the method `referential` exists for struct `Selfie<'_, UnstableInt, selfie::refs::Ref<i32>>`, but its trait bounds were not satisfied
   --> tests/compile_fail/unstable_deref.rs:20:27
    |
6   | struct UnstableInt(pub i32);
    | ---------------------------- doesn't satisfy `UnstableInt: stable_deref_trait::StableDeref`
...
20  |     assert_eq!(&42, *data.referential());
    |                           ^^^^^^^^^^^ private field, not a method
    |
    = note: the following trait bounds were not satisfied:
            `UnstableInt: stable_deref_trait::StableDeref`
note: the following trait must be implemented
   --> $CARGO/stable_deref_trait-1.2.0/src/lib.rs
    |
    | pub unsafe trait StableDeref: Deref {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: the method `referential` exists for struct `Box<Selfie<'_, UnstableInt, selfie::refs::Ref<i32>>>`, but its trait bounds were not satisfied
   --> tests/compile_fail/unstable_deref.rs:23:27
    |
6   | struct UnstableInt(pub i32);
    | ---------------------------- doesn't satisfy `UnstableInt: stable_deref_trait::StableDeref`
...
23  |     assert_eq!(&42, *data.referential());
    |                           ^^^^^^^^^^^ private field, not a method
    |
    = note: the following trait bounds were not satisfied:
            `UnstableInt: stable_deref_trait::StableDeref`
note: the following trait must be implemented
   --> $CARGO/stable_deref_trait-1.2.0/src/lib.rs
    |
    | pub unsafe trait StableDeref: Deref {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
